### ISSUE TITLE
MergeDeep concatenates lists instead of replacing them

### ISSUE THREAD

**Issue by timothyjlaurent** [immutable-js/immutable-js#1452]
### What happened

When mergingDeep Lists are concatenated instead of replacing the list at the key.

### How to reproduce

```
Immutable.fromJS({
  list: [1,2]
}).mergeDeep(Immutable.fromJS({
  list: [1,2]
})).toJS()
```

returns  `{list: [1,2,1,2]}`

expect `{list: [1,2]}`

`mergeDeepWith` is no help as Lists are skipped by the resolver as mentioned in #1420

**Comment by timothyjlaurent** [comment]
This works in V3x.

**Comment by timothyjlaurent** [comment]
My preference for the expected behavior is this:

When a nested list is merged deep the merge should happen recursively at each index of the lists:

```
expect(Immutable.fromJS({
  list: [1,{a: 1}, 3]
}).mergeDeep(Immutable.fromJS({
  list: [-1,{b: 2}]
})).toJS()).toEqual({
  list: [-1, {a:1, b:2}]
})
```

If we want to patch a list at an index, we could potentially pass in another keyed collection with keys at the indexes we'd like to merge:

```
expect(Immutable.fromJS({
  list: [1,{a: 1}, 3]
}).mergeDeep(Immutable.fromJS({
  list: Immutable.map({1: -1})
})).toJS()).toEqual({
  list: [-1, {a:1}, 3]
})
``` 

What do you think of that @leebyron ^^

**Comment by JanEggers** [comment]
any updates on this one?

im using 4.0.0 rc9 and have the same issue with merge!

**Comment by thessem** [comment]
This behaviour changed somewhere between rc4 and rc9

**Comment by acusti** [comment]
This change is described in the [release notes](https://github.com/facebook/immutable-js/releases/tag/v4.0.0-rc.5) for 4.0.0-rc.5 and landed in PR #1344.

In practical terms, we never want this new behavior in our application, so we created two utils to handle our particular use cases. `mergeDeep` does the old-style mergeDeep where Lists are deep merged at matching indexes. `mergeDeepOverwriteLists`, as the name suggests, will overwrite any existing List with the List being merged into it. Here they are, along with 


```js
const isMergeable = (a) => (
    a && typeof a === 'object' && typeof a.mergeWith === 'function' && !List.isList(a)
);

export const mergeDeep = (a, b) => {
    // If b is null, it would overwrite a, even if a is mergeable
    if (isMergeable(a) && b !== null) {
        return a.mergeWith(mergeDeep, b);
    }

    if (!List.isList(a) || !List.isList(b)) {
        return b;
    }

    return b.reduce((acc, nextItem, index) => {
        const existingItem = acc.get(index);
        if (isMergeable(existingItem)) {
            return acc.set(index, existingItem.mergeWith(mergeDeep, nextItem));
        }

        return acc.set(index, nextItem);
    }, a);
};

export const mergeDeepOverwriteLists = (a, b) => {
    // If b is null, it would overwrite a, even if a is mergeable
    if (b === null) return b;

    if (isMergeable(a) && !List.isList(a)) {
        return a.mergeWith(mergeDeepOverwriteLists, b);
    }

    return b;
};
```

**Comment by abrcdf1023** [comment]
In case someone is looking for another solution.
https://github.com/immutable-js/immutable-js/issues/762
